### ISSUE TITLE
Gstreamer replaygain test erroring if gstreamer installed, but gst-plugins-good not installed

### ISSUE THREAD

**Issue by diego-plan9** [beetbox/beets#1830]
A (minor and system-specific) issue found while running the full test suite on my environment: it seems that if `gstreamer` is installed but `gst-plugins-good` (which contains the `rganalysis` plugin) is not installed, `test_replaygain.ReplayGainCmdCliTest` **is** run, and its tests result in an error (during their `setUp()`) due to the absence of the mentioned plugin: 

```
======================================================================
ERROR: test_cli_saves_album_gain_to_file (test_replaygain.ReplayGainGstCliTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/debian/d3/beets-1.3.16+dfsg/test/test_replaygain.py", line 49, in setUp
    self.load_plugins('replaygain')
  File "/tmp/debian/d3/beets-1.3.16+dfsg/test/helper.py", line 207, in load_plugins
    beets.plugins.find_plugins()
  File "/tmp/debian/d3/beets-1.3.16+dfsg/beets/plugins.py", line 289, in find_plugins
    _instances[cls] = cls()
  File "../beetsplug/replaygain.py", line 817, in __init__
    'replaygain initialization failed: {0}'.format(e)
UserError: replaygain initialization failed: no replaygain command found: install mp3gain or aacgain
...
```

I believe it takes a rather unlikely system configuration to trigger this problem, but it's not entirely out of the blue: for example, [python-gst0.10 on Debian](https://packages.debian.org/sid/python-gst0.10) does not depend on `gst-plugins-good` (even if it usually ends up being pulled into the system by other packages anyway).

While this issue seems to have been addressed on #1518 at the execution level, I'm wondering if it would be a good idea to tweak `test_replaygain.ReplayGainCmdCliTest` a little as well in order to skip the test unless both gstreamer **and** the required plugins are installed?

**Comment by sampsyo** [comment]
Good catch. The test should definitely be skipped if the necessary components are not available. I can imagine doing this either by expanding that `GST_AVAILABLE` check inline or by reusing the code from the plugin that checks for the backend.

Weirdly, I also note that this error message is incorrect:

> replaygain initialization failed: no replaygain command found: install mp3gain or aacgain

Since we're in `ReplayGainGstCliTest`, we shouldn't see the message from `CommandBackend`! I don't see what's going wrong here.

**Comment by diego-plan9** [comment]
> Since we're in ReplayGainGstCliTest, we shouldn't see the message from CommandBackend! I don't see what's going wrong here.

After some checking, I believe it is due to the call order in `ReplayGainCliTestBase.setUp()`: `config['replaygain']['backend']` is initialized *after* `self.load_plugins('replaygain')`, and as a result `CommandBackend` is always the backend that is initialized?

**Comment by sampsyo** [comment]
You were absolutely right -- and we weren't testing what we thought we were testing. I moved the configuration line and it seems to be using the right backend now.

**Comment by diego-plan9** [comment]
Thanks for the quick fix!

I have been playing with the tests in order to solve the initial issue, with moderate luck: I don't really have experience with `gstreamer`, and it would be great if you or other more experienced developer could double-check and provide input.

Basically:
* would it cause problems with `gstreamer` to simply create a GStreamerBackend, along the lines of
```python
try:
    import gi
    gi.require_version('Gst', '1.0')
    GST_AVAILABLE = True

    # Check for plugins as well by initializing a GStreamerBackend.
    gstreamer_backend = GStreamerBackend(config['replaygain'], None)
except (ImportError, ValueError, FatalReplayGainError):
    GST_AVAILABLE = False
```
? My main doubt is that, if using this approach, things like `GObject.threads_init()` and `Gst.init([sys.argv[0]])` get called twice - which I'm not sure if it safe or supported.


* Also, I have run into a problem (only on `test_cli_saves_track_gain`) caused at runtime, and *as far I can see* hints towards another pluging missing (a MP3 decoder?, which [seems to only be provided](http://gstreamer.freedesktop.org/documentation/plugins.html) by `gst-plugins-bad` (mpg123) or `gst-plugins-ugly` (mad)) :

```
============================= ERRORS =============================
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/diego/venvs/beets/git/beets/test/test_replaygain.py", line 97, in test_cli_saves_track_gain
    self.assertIsNotNone(item.rg_track_peak)
  File "/usr/lib/python2.7/unittest/case.py", line 960, in assertIsNotNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: unexpectedly None
-------------------- >> begin captured logging << --------------------
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: item_copied
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: item_copied
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: write
beets: DEBUG: Sending event: after_write
beets: DEBUG: Sending event: database_change
beets: DEBUG: Sending event: write
beets: DEBUG: Sending event: after_write
beets: DEBUG: Sending event: database_change
beets: DEBUG: no user configuration found at /tmp/tmpuy58bU/config.yaml
beets: DEBUG: data directory: /tmp/tmpuy58bU
beets: DEBUG: plugin paths: 
beets: DEBUG: Sending event: pluginload
beets.replaygain: INFO: replaygain: analyzing the artist - älbum - tïtle 0
beets.replaygain: INFO: replaygain: ReplayGain error: Error GLib.Error('Your GStreamer installation is missing a plug-in.', 'gst-core-error-quark', 12) - 'gstdecodebin2.c(3977): gst_decode_bin_expose (): /GstPipeline:pipeline1/GstDecodeBin:decbin:\nno suitable plugins found' on file '/tmp/tmpuy58bU/libdir/the artist/\xc3\xa4lbum/02 t\xc3\xaftle 0.mp3'
beets.replaygain: INFO: replaygain: analyzing the artist - älbum - tïtle 1
beets.replaygain: INFO: replaygain: ReplayGain error: Error GLib.Error('Your GStreamer installation is missing a plug-in.', 'gst-core-error-quark', 12) - 'gstdecodebin2.c(3977): gst_decode_bin_expose (): /GstPipeline:pipeline1/GstDecodeBin:decbin:\nno suitable plugins found' on file '/tmp/tmpuy58bU/libdir/the artist/\xc3\xa4lbum/02 t\xc3\xaftle 1.mp3'
beets: DEBUG: Sending event: cli_exit
--------------------- >> end captured logging << ---------------------
```

Again, this might only happen in [very specific configurations](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=742508), and maybe not worth the effort for including an extra check for the existence of the decoder. However, perhaps it's trivial to find out if a plugin is installed (and/or if a specific format is supporter) using the `gstreamer` API - I'd be happy to devote some time checking their docs eventually if no one steps up!

**Comment by sampsyo** [comment]
I'm pretty sure temporarily instantiating a `GStreamerBackend` should be fine; it shouldn't be a problem to call those twice. (But maybe that code should be in a `setUp` method instead of top-level in the module...)

And that's a frustrating failure. It seems like we probably should tolerate the case where decoders are missing. For the purpose of the tests, though, we could consider just skipping the test if the result is None (instead of failing)---assuming that the only way it could be None is that a decoder is missing.

Testing this gstreamer code is a messy business.

**Comment by jtpavlock** [comment]
closed per #1836
